source/fuzz/fuzzer_pass_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,164 @@
+        // Only run fuzzer pass on valid arithmetic operation instruction.
+        if (!valid_arithmetic_types.count(instruction_iterator->opcode()))
+          return;
+
+        assert(instruction_iterator->opcode() ==
+                   instruction_descriptor.target_instruction_opcode() &&
+               "The opcode of the instruction we might insert before must be "
+               "the same as the opcode in the descriptor for the instruction");
+
+        // Randomly decide whether to wrap it to a vector operation.
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfWrappingVectorSynonym())) {
+          return;
+        }
+
+        // It must be valid to insert an OpCompositeConstruct instruction
+        // before |instruction_iterator|.
+        if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(
+                SpvOpCompositeConstruct, instruction_iterator)) {
+          return;
+        }
+        // Get the scalar type represented by the targeted instruction id.
+        uint32_t scalar_type_id = instruction_iterator->type_id();
+
+        // Get a random vector size from 2 to 4.
+        uint32_t component_count = 2 + std::rand() % 3;
+        // Get the vector type of size range from 2 to 4 of the corresponding
+        // scalar type.
+        uint32_t vec_type_id =
+            FindOrCreateVectorType(scalar_type_id, component_count);
+
+        // Randomly choose a position that target ids should be placed at.
+        // The position is in range [0, n - 1], where n is the size of the
+        // vector.
+        uint32_t position = std::rand() % component_count;
+
+        // target ids are the two scalar ids from the original instruction.
+        uint32_t target_id1 = instruction_iterator->GetSingleWordOperand(1);
+        uint32_t target_id2 = instruction_iterator->GetSingleWordOperand(2);
+
+        // Stores the ids of scalar constants.
+        std::vector<uint32_t> vec1_components;
+        std::vector<uint32_t> vec2_components;
+
+        // Width is specified in the first index for either OpTypeInt or
+        // OpTypeFloat.
+        uint32_t width = GetIRContext()
+                             ->get_def_use_mgr()
+                             ->GetDef(scalar_type_id)
+                             ->GetSingleWordOperand(0);
+        // Get the scalar type.
+        auto type = GetIRContext()->get_type_mgr()->GetType(scalar_type_id);
+
+        // Whether the constant is signed, not used for float type so default
+        // set to true.
+        bool is_signed =
+            type->AsInteger() ? type->AsInteger()->IsSigned() : true;
+
+        // Populate components based on vector type and size.
+        for (uint32_t i = 0; i < component_count; ++i) {
+          if (i == position) {
+            vec1_components.emplace_back(target_id1);
+            vec2_components.emplace_back(target_id2);
+          } else {
+            switch (instruction_iterator->opcode()) {
```

Comment:
`instruction_iterator->opcode()` is the opcode of the instruction (e.g. `OpIAdd`), not the type of the instruction.

It might be easier to do different things depending on the `type` as follows:
```cpp
if (const auto* int_type = type->AsInteger()) {
  // operands are integers
} else if (const auto* float_type = type->AsFloat()) {
  // operands are floats
}
```

Reply:
Done.